source/simulator/entropy_viscosity.cc

```diff
@@ -319,6 +325,81 @@ namespace aspect
+                  if (tangential_velocity_boundaries.find(cell->face(face_no)->boundary_id()) != tangential_velocity_boundaries.end() ||
+                      zero_velocity_boundaries.find(cell->face(face_no)->boundary_id()) != zero_velocity_boundaries.end())
+                    {
+                      cell_at_conduction_dominated_dirichlet_boundary = true;
+                      continue;
+                    }
+
+                  scratch.face_finite_element_values->reinit (cell, face_no);
+                  (*scratch.face_finite_element_values)[introspection.extractors.velocities].get_function_values(old_solution,
+                      face_old_velocity_values);
+                  (*scratch.face_finite_element_values)[introspection.extractors.velocities].get_function_values(old_old_solution,
+                      face_old_old_velocity_values);
+
+                  // ... check if the face is a boundary with normal flow by integrating the normal velocities
+                  // (flux through the boundary) as: int u*n ds = Sum_q u(x_q)*n(x_q) JxW(x_q)...
+                  double normal_flow = 0.0;
+                  double flow = 0.0;
+                  double area = 0.0;
+                  for (unsigned int q=0; q<scratch.face_finite_element_values->n_quadrature_points; ++q)
+                    {
+                      normal_flow += ((face_old_velocity_values[q]+face_old_old_velocity_values[q])/2.0 *
+                                      scratch.face_finite_element_values->normal_vector(q)) *
+                                     scratch.face_finite_element_values->JxW(q);
+                      flow += ((face_old_velocity_values[q]+face_old_old_velocity_values[q])/2.0).norm() *
+                              scratch.face_finite_element_values->JxW(q);
+                      area += scratch.face_finite_element_values->JxW(q);
+                    }
+
+                  // Disable stabilization for boundaries with slow flow, or tangential flow.
+                  // Break the loop in case a face is at multiple boundaries, some with flow, some without.
+                  // In those cases we can not disable stabilization.
+                  if ((std::abs(flow/area) < std::sqrt(std::numeric_limits<double>::epsilon()) * area) ||
+                      std::abs(normal_flow) < std::sqrt(std::numeric_limits<double>::epsilon()) * std::abs(flow))
+                    {
+                      cell_at_conduction_dominated_dirichlet_boundary = true;
+                    }
+                  else
+                    {
+                      cell_at_conduction_dominated_dirichlet_boundary = false;
+                      break;
+                    }
+                }
+
+            if (cell_at_conduction_dominated_dirichlet_boundary)
+              {
+                viscosity_per_cell[cell->active_cell_index()] = 0.0;
+                continue;
```

Comment:
add comment:
// if viscosity is zero, we don't need any further computation on this cell